File: backend/app/middleware/auth_middleware.py
from functools import wraps
from flask import request, jsonify
from app.services.auth_service import verify_token


def jwt_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):

        auth_header = request.headers.get("Authorization")

        if not auth_header:
            return jsonify({"error": "Token missing"}), 401

        token = auth_header.split(" ")[1]

        decoded = verify_token(token)

        if not decoded:
            return jsonify({"error": "Invalid or expired token"}), 401

        request.user = decoded

        return f(*args, **kwargs)

    return decorated


def role_required(role):
    def decorator(f):

        @wraps(f)
        def decorated(*args, **kwargs):

            if request.user["role"] != role:
                return jsonify({"error": "Unauthorized"}), 403

            return f(*args, **kwargs)

        return decorated

    return decorator

File: backend-api/middleware/authMiddleware.js
const jwt = require("jsonwebtoken");

const JWT_SECRET = process.env.JWT_SECRET;


// Verify JWT token
exports.verifyToken = (req, res, next) => {

    const authHeader = req.headers["authorization"];

    if (!authHeader) {
        return res.status(401).json({ error: "Token required" });
    }

    const token = authHeader.split(" ")[1];

    try {

        const decoded = jwt.verify(token, JWT_SECRET);

        req.user = decoded;

        next();

    } catch (err) {
        return res.status(401).json({ error: "Invalid token" });
    }
};


// Role based access control
exports.requireRole = (role) => {

    return (req, res, next) => {

        if (!req.user) {
            return res.status(401).json({ error: "Unauthorized" });
        }

        if (req.user.role !== role) {
            return res.status(403).json({ error: "Access denied" });
        }

        next();
    };
};